src/process_watch.js

```diff
@@ -0,0 +1,369 @@
+      return element.pid == processObject.pid
+    })
+    /* 
+     * found time, Show how long a ps has been on list
+     * age should be difference between found and current time 
+     */
+    processObject.firstDetectTime = processWarnList[index].firstDetectTime;
+    // reportTime, Show how long since last report
+    processObject.reportTime = processWarnList[index].reportTime;
+    // If a new process, that is repeat offender, allow reporting it
+    // Also a little over 15 min old 
+    processObject.report = ((processObject.reportTime == undefined) && (
+      processObject.firstDetectTime <= (currentTime - 15 * 60 * 1000)));
+    //overwrite previous
+    processWarnList.splice(index, 1, processObject);
+  } else {
+    processObject.lastDetectTime = currentTime;
+    processObject.firstDetectTime = currentTime;
+    processObject.report = false;
+    processWarnList.push(processObject);
+  }
+}
+
+module.exports = {
+  /**
+   * init
+   * @param {ps} current-processes npm module
+   * @param {pid} pidusage npm module
+   * @param {steam} output stream for reports. Usually Slack. STDout for testing
+   * bring in the process watching modules
+   * AND the stream for output to Slack
+   */
+  init: function (ps, pid, stream) {
+    psLookup = ps;
+    pidusage = pid;
+    currStream = stream;
+  },
+
+  /**
+   * checkProcess
+   * @param {currentTime} used track in lists when 'new' and when 'updated'
+   * On first new run we check cpu type
+   * 
+   * new cpuAddLevel processes are added to Watch list
+   * If a  PID 'match' is in Watch already, process is added to Warn list
+   * 
+   * If tracked process is above cpuAddLevel, it's 'last detected' timestamp is updated
+   * If a process is below cpuAddLevel, it can eventually age out of the list
+   * 
+   * 'cleanList' is then run with 'processes,' a current snapshot of all processes.
+   */
+  checkProcess: function checkProcess(currentTime = new Date().getTime()) {
+    // stops function from crashing server if init hasn't been run
+    if (!psLookup) {
+      console.log("process_watch not initialized");
+      return null;
+    }
+    cpuTypeChecked = false;
+    psLookup.get(function (err, processes) {
+
+      var sorted = _.sortBy(processes, "cpu");
+      let top5 = sorted.reverse().splice(0, 5);
+      top5.forEach(element => {
+        // tests need to match what OS returns for cpu
+        // 25 or .25
+        if (!cpuTypeChecked) {
+          updateCPUTestValues(element.cpu);
+        }
+        //  This allows testing enviroment to pass in times,
+        //  normally current-processes doesn't pass a timestamp
+        currentTime = ((typeof (element.timestamp) !==
+          'undefined') ? element.timestamp : currentTime);
+
+        // check if high CPU offender before running log updates
+        if (element.cpu >= cpuAddLevel) {
+          var match = function (stored) {
+            if (stored === undefined) {
```

Comment:
huh!  I have never used this one before.  I've always used `typeof X === "undefined"`.

You learn something new everyday!

_( just make sure you use this on a local variable you have declared, or else this with throw an error )_

Reply:
Yes, this is a weird case. I think this is to check that the function can deal with passed null data. maybe if processWatchList is actually empty? But then why is .some() running on an empty array? 

This was a fast patch to make the code run, I'm sure there's a more elegant solution